Tests/DependenciesTests/DependencyValuesTests.swift

```diff
@@ -692,8 +692,58 @@ final class DependencyValuesTests: XCTestCase {
     @Dependency(\.date) var date
     _ = date
   }
+  
+  func testNoDeadlock() {
+    let expectation = XCTestExpectation(description: "Should be fulfilled on init")
+    
+    DispatchQueue.global(qos: .default).async {
+      @Dependency(\.dependencyWithSyncDependency) var dependency
+      XCTAssertGreaterThanOrEqual(dependency.count, 0)
+      expectation.fulfill()
+    }
+    
+    wait(for: [expectation], timeout: 5)
+  }
+}
+
+// MARK: Deadlock Example
+
+struct DependencyWithSyncDependency: TestDependencyKey {
+  let count: Int
+  
+  init() {
+    // This is a very contrived example. Initializing this dependency
+    // depends on an `@Dependency()` call that happens in another thread. Because
+    // this initializer does not complete until it gets the value from the other
+    // thread, it holds the `CachedValues.lock` and prevents the other `@Dependency`
+    // call from ever returning.
+    let group = DispatchGroup()
+    group.enter()
+    var count = -1
+    
+    DispatchQueue.global(qos: .userInitiated).async {
+      @Dependency(\.countInitDependency) var countInitDependency
+      count = countInitDependency.fetch()
+      group.leave()
+    }
+    
+    group.wait()
```

Comment:
This all just simulates waiting for an `@Dependency` on another thread. Using `DispatchQueue.sync` would be easier to read, but within the test that actually reuses the `init` thread preventing the issue. Using `async` with `group.wait()` forces an actual thread change.